go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -335,12 +342,192 @@ func (c *Controller) reconcileCUDN(key string) error {
 	return errors.Join(syncErr, updateStatusErr)
 }
 
-func (c *Controller) syncClusterUDN(_ *cudnv1.ClusterUserDefinedNetwork) ([]netv1.NetworkAttachmentDefinition, error) {
-	// TODO: implement
-	return nil, fmt.Errorf("implement me")
+func (c *Controller) syncClusterUDN(cudn *cudnv1.ClusterUserDefinedNetwork) ([]netv1.NetworkAttachmentDefinition, error) {
+	if cudn == nil {
+		return nil, nil
+	}
+
+	selectedNamespaces, err := c.getSelectedNamespaces(cudn.Spec.NamespaceSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get selected namespaces: %w", err)
+	}
+	affectedNamespaces := c.namespaceTracker[cudn.Name]
+
+	if !cudn.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(cudn, template.FinalizerUserDefinedNetwork) {
+			var errs []error
+			for nsToDelete := range affectedNamespaces.Union(selectedNamespaces) {
+				if err := c.deleteNAD(cudn, nsToDelete); err != nil {
+					errs = append(errs, fmt.Errorf("failed to delete NetworkAttachmentDefinition: %w", err))
+				}
+				c.namespaceTracker[cudn.Name].Delete(nsToDelete)
```

Comment:
should we only do that if NAD deletion succeeded?

Reply:
Fixed, thanks!

Reply:
Fixed by last push https://github.com/ovn-org/ovn-kubernetes/pull/4612#issuecomment-2353034446